### DISCUSSION TITLE
Use template in custom handler for 301 redirect

### DISCUSSION THREAD

**Discussion by suntong** [fastapi/fastapi#6194] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI, Request, Response, Form, status
from fastapi.templating import Jinja2Templates

app = FastAPI()
templates = Jinja2Templates(directory='templates')

@app.get("/redirect")
async def redirectu(request: Request, url: str):
    headers = {'Location': url}
    return templates.TemplateResponse('redirect_body.html', {'request': request, 'url': url})
```

### Description

Similar to issue #3182, I want to use the Jinja2 template to render my custom template using TemplateResponse for _301 redirect_ response. 

I got the above code work fine. However, it is missing two critical parts as 301 redirect response:

- the http return code is 200 instead of 301, and
- the `headers = {'Location': url}` is not generating the proper 301 redirect headers. 

How can I make the above two changes to my code?

More details:

The HTTP Header I got from above code is:

```
< HTTP/1.1 200 OK
< Date: Wed, 16 Nov 2022 22:31:19 GMT
< Content-Type: text/html; charset=utf-8
```

Whereas I want my HTTP Header to be:

```
< HTTP/2 301 
< content-type: text/html; charset=UTF-8
< location: url passed from get
```

PS. 

I've also tried:

    return RedirectResponse(url=url, status_code=301)

but it is missing the important custom template part of the http body.

PPS. 

Here is a proper 301 redirect example (the kind of return we get from tinyurl):

```
< HTTP/2 301 
< content-type: text/html; charset=UTF-8
< location: https://github.com/vuejs/vue.git
<!DOCTYPE html>
<html>
    <head>
        <meta charset="UTF-8" />
        <meta http-equiv="refresh" content="0;url='https://github.com/vuejs/vue.git'" />
        <title>Redirecting to https://github.com/vuejs/vue.git</title>
    </head>
    <body>
        Redirecting to <a href="https://github.com/vuejs/vue.git">https://github.com/vuejs/vue.git</a>.
    </body>
</html>
```

### Operating System

Linux

### Operating System Details

The code is _actually_ hosted in deta micro:

Collecting fastapi
https://files.pythonhosted.org/packages/8c/f5/68a7a099ae3674e85e2de11d3859ebffd7fe32a8cda0d11bd4d43f32437b/fastapi-0.87.0-py3-none-any.whl
Collecting jinja2
  Downloading https://files.pythonhosted.org/packages/bc/c3/f068337a370801f372f2f8f6bad74a5c140f6fda3d9de154052708dd3c65/Jinja2-3.1.2-py3-none-any.whl (133kB)

I.e., I have fastapi-0.87.0 & jinja2-3.1.2 

But I'm not too sure about the python version.

### FastAPI Version

fastapi-0.87.0

### Python Version

3.x

### Additional Context

_N/A_

**Comment by jgould22** [comment, marked as answer]
You are just returning a template, which is a 200 by default.

You need to tell FastAPI that the response is a 301 and give it a header (how else is it to know?)

Try 
```python
    return templates.TemplateResponse('redirect_body.html', {'request': request, 'url': url}, headers={'Location': url}, status_code=301)
```

You can see the function definition that you are calling here https://github.com/encode/starlette/blob/master/starlette/templating.py#L86

**Comment by suntong** [comment]
Bravo! Thanks a lot!!

**Comment by tiangolo** [comment]
Thanks for the help here @jgould22 ! 👏  🙇 

Thanks for reporting back and closing the issue @suntong  👍
